packages/core/src/agent/EnvelopeService.ts

```diff
@@ -29,38 +30,44 @@ class EnvelopeService {
   }
 
   public async packMessage(payload: AgentMessage, keys: EnvelopeKeys): Promise<EncryptedMessage> {
-    const { routingKeys, senderKey } = keys
-    let recipientKeys = keys.recipientKeys
+    const { recipientKeys, routingKeys, senderKey } = keys
+    let recipientVerkeys = recipientKeys.map(didKeyToVerkey)
+    const routingVerkeys = routingKeys.map(didKeyToVerkey)
+    const senderVerkey = senderKey && didKeyToVerkey(senderKey)
 
     // pass whether we want to use legacy did sov prefix
     const message = payload.toJSON({ useLegacyDidSovPrefix: this.config.useLegacyDidSovPrefix })
 
     this.logger.debug(`Pack outbound message ${message['@type']}`)
 
-    let encryptedMessage = await this.wallet.pack(message, recipientKeys, senderKey ?? undefined)
+    let encryptedMessage = await this.wallet.pack(message, recipientVerkeys, senderVerkey ?? undefined)
 
     // If the message has routing keys (mediator) pack for each mediator
-    for (const routingKey of routingKeys) {
+    for (const routingVerkey of routingVerkeys) {
       const forwardMessage = new ForwardMessage({
         // Forward to first recipient key
-        to: recipientKeys[0],
+        to: recipientVerkeys[0],
         message: encryptedMessage,
       })
-      recipientKeys = [routingKey]
+      recipientVerkeys = [routingVerkey]
       this.logger.debug('Forward message created', forwardMessage)
 
       const forwardJson = forwardMessage.toJSON({ useLegacyDidSovPrefix: this.config.useLegacyDidSovPrefix })
 
       // Forward messages are anon packed
-      encryptedMessage = await this.wallet.pack(forwardJson, [routingKey], undefined)
+      encryptedMessage = await this.wallet.pack(forwardJson, [routingVerkey], undefined)
     }
 
     return encryptedMessage
   }
 
   public async unpackMessage(encryptedMessage: EncryptedMessage): Promise<DecryptedMessageContext> {
-    return this.wallet.unpack(encryptedMessage)
+    const decryptedMessage = await this.wallet.unpack(encryptedMessage)
+    const { recipientKey, senderKey, plaintextMessage } = decryptedMessage
```

Comment:
Not necessarily for now, but what if we instead start passing around `Key` instances instead of the string? I think that may help with all the transformations we need to do with `verkeyToDidKey` and `didKeyToVerkey`? If we just add one helper method on the didcomm service to access the actual key instances

Reply:
What do you think of this idea @jakubkoci? I quite like it, and as we're using `Key` in more and more places it starts to make more sense.

e.g. in this PR we update the wallet interface to use `Key` instances instead of leveraging `verkeys` : https://github.com/hyperledger/aries-framework-javascript/pull/684/files#diff-3789ea30a1c820026da4fd9d886aa730250fc2a99a8dc62bb6c3143f8b798aefR41-R56